### ISSUE TITLE
Alembic support

### ISSUE THREAD

**Issue by gerroon** [f3d-app/f3d#138]
**Describe the solution you'd like**
It would be very nice if this app also supports alembic/alembic caches

**Describe alternatives you've considered**
Open the Alembic files in Maya 

**Additional context**

**Contribution**
I am not a developer.

**Comment by Meakk** [comment]
Indeed, I think the format is very popular and f3d should support it.  
Do you have some files to share so we can test and validate the implementation when we'll do it?

**Comment by mwestphal** [comment]
For reference:

https://github.com/alembic/alembic

https://github.com/assimp/assimp/issues/1835

**Comment by gerroon** [comment]
> Indeed, I think the format is very popular and f3d should support it. Do you have some files to share so we can test and validate the implementation when we'll do it?

Hi,

Here are two trees without textures, you can find the actual set in the link below. I would be happy to generate larger alembic files if you need.

https://www.artstation.com/marketplace/p/qbdBm/free-beech-tree-set

**Comment by nyue** [comment]
If the Alembic file I/O development requires starting from scratch, where should the effort be put into i.e. adding it to ASSIMP or adding it to VTK ?

**Comment by Meakk** [comment]
VTK imho. @mwestphal and I have extensive experience with VTK so we can help. @mwestphal is also maintainer of VTK.

**Comment by nyue** [comment]
@Meakk which existing VTK class should I base my initial study on that may help me in my experimentation ?

**Comment by mwestphal** [comment]
vtkGLTFImporter is the most modern importers in VTK, with all features, although a bit complex.

vtkOBJImporter is the reliable and simple one.

**Comment by nyue** [comment]
Thank you, I shall start looking at the OBJ as an example

For OBJ, I see

- vtkOBJImporter(vtkImporter)
- vtkOBJExporter(vtkExporter)
- vtkOBJReader(vtkAbstractPolyDataReader)
- vtkOBJWriter(vtkWriter)

For Alembic in the long run, should one be keeping the above at the back of our mind when designing and coding ? i.e. there may be

- vtkABCImporter(vtkImporter)
- vtkABCExporter(vtkExporter)
- vtkABCReader(vtkAbstractPolyDataReader)
- vtkABCWriter(vtkWriter)

In the film/vfx industry, sometimes we refer to Alembic files as ABC because of it's extension, in terms of naming convention in VTK, should I name the new class as vtkAlembic* or vtkABC* ?

Cheers

**Comment by Meakk** [comment]
@nyue you will need only `vtkABCImporter` for F3D. Using `ABC` prefix is fine.

**Comment by mwestphal** [comment]
I'd rather use Alembic prefix, much more clear for someone not knowing what it is.

Only the importer is needed.

**Comment by nyue** [comment]
I have an initial working prototype (just plain geometry, not uvs, normals, colors etc). I guess now I following the steps [here](https://gitlab.kitware.com/vtk/vtk/blob/master/Documentation/dev/git/develop.md) ?

**Comment by Meakk** [comment]
@nyue yes. Do you rely on an external library?

**Comment by mwestphal** [comment]
@nyue Yes, please tag me on the MR so I can help you with the needed steps :)

**Comment by nyue** [comment]
> @nyue yes. Do you rely on an external library?

Yes. [Alembic](https://github.com/alembic/alembic) which depends on [OpenEXR](https://github.com/AcademySoftwareFoundation/openexr) or more specifically the [Imath](https://github.com/AcademySoftwareFoundation/Imath) part.

In the past, I have build and use all of them on the major platforms. Windows, Linux and OS X (x86_64)

I have also recently been building and using them on (ARM_64)

Cheers

**Comment by Meakk** [comment]
Then you may have difficulty to include it into VTK, what do you think @mwestphal ?  
Alternatively, you can include the `vtkAlembicImporter` directly in F3D and use an optional `find_package(Alembic)` in CMake depending on a new option `F3D_MODULE_ALEMBIC`.

**Comment by mwestphal** [comment]
I definitely agree. You can take a look at the Assimp reader integration as an inspiration.

**Comment by nyue** [comment]
@Meakk noted.

@mwestphal OK, I will have a look at Assimp in context of f3d.

**Comment by nyue** [comment]
I have a working prototype on my fork of f3d

https://github.com/nyue/f3d?organization=nyue&organization=nyue

Not sure what is the development/review process. Hopefully it is a simple.

**Comment by Meakk** [comment]
@nyue this looks really good!  
You can open a PR when you are ready, and @mwestphal or I will review the code and test the functionality.  
I'm curious about the Alembic format: Is it supporting cameras, lights, texture materials?

**Comment by gerroon** [comment]
> I'm curious about the Alembic format: Is it supporting cameras, lights, texture materials?

Yes, it also has vertex cache animation support. It is different than FBX, no bones or soft deformations.

**Comment by nyue** [comment]
Thanks for chiming in @gerroon , what DCC are you using to generate the Alembics? Currently the code supports simple geometry (no normals, uvs and ArbGeomParam) and basic transform. In your intended use cases, what features are a priority for you ?

Cheers

**Comment by gerroon** [comment]
@nyue 

I use Blender and Houdini. 

Thanks for implementing the initial version. However to make it minimally useful it would be nice to be able to have uvs, normals and finally the cache playback support. I think at this point the arbitrary parameters would not be too useful given they could be DCC specific. 

It would also be nice if it can at least load the initial assets if the .abc file is a animated cache instead  of a static asset.

**Comment by nyue** [comment]
@gerroon Thanks for the background. I am also using Blender but mostly Houdini.

I am new to f3d so I will need to understand the design of f3d in terms of time sampling to be able to implement the cache playback.

Getting the initial PR accepted into the main repo is important (this ensures that my code play nice with GitHub CI/CD and the dependency [Alembic] are sorted out in GitHub CI/CD builds)

Once the initial PR is merged, I will incrementally add normals, UVs as they are low hanging fruits before I tackle cache playback.

Cheers

**Comment by gerroon** [comment]
@nyue 

Thanks for the hard work, Alembic support is going to be super nice given there are not really cross platform apps to see it. F3D can be the ultimate 3D browser once the VDB is also implemented at some point.

**Comment by nyue** [comment]
VTK itself has VDB IO

https://vtk.org/doc/nightly/html/dir_f71ab6aedd1615291b1af0dba141160d.html

**Comment by mwestphal** [comment]
VTK only contains a vtkVDBWriter, no vtkVDBReader, but it would definitelly make sense to add one. This request should go into its own issue though.

**Comment by mwestphal** [comment]
This issue is fixed has we now have partial Alembic support ! thanks @nyue !
